[KAFKA-3729] Auto-configure non-default SerDes passed alongside the topology builder

tedyu opened the pull request:
Only default serdes provided through configs are auto-configured today. But we could auto-configure other serdes passed alongside the topology builder as well.

Added new tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

tedyu committed:
[KAFKA-3729] Auto-configure non-default SerDes passed alongside the topology builder

tedyu committed:
Address checkstyle

tedyu committed:
handle null reference

tedyu committed:
handle null reference

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -634,6 +636,25 @@ private KafkaStreams(final InternalTopologyBuilder internalTopologyBuilder,
+                sn.getValueDeSer().configure(config.originals(), false);
+            }
+        }
+    }
```
nit: empty line after 657 here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredWindowStore.java:
```diff
@@ -99,6 +99,12 @@ public void init(final ProcessorContext context,
 
     @SuppressWarnings("unchecked")
     void initStoreSerde(final ProcessorContext context) {
+        if (keySerde != null) {
```
Why we would not need to check `context.appConfigs() != null` here?

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/integration/KStreamAggregationDedupIntegrationTest.java:
```diff
@@ -170,16 +180,79 @@ public void shouldReduceWindowed() throws Exception {
+            if (!((MyIntDeserializer) this.deserializer()).configured()) return false;
+            return true;
+        }
+    }
```
nit: empty line.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -476,7 +476,63 @@ public void shouldNotGetTaskWithKeyAndPartitionerWhenNotRunning() {
+        }
+
+        boolean configured() {
+            return !called || configured;
```
Why we want to validate that deserialize is not called while configure is called? Ditto else where.

tedyu replied:
The above condition should be read like this:

the serde should be configured (configured == true), otherwise the ser(de) is not used during the test.
This allows the symmetrical check shown below:
```
            if (!((MyIntSerializer) this.serializer()).configured()) return false;
            if (!((MyIntDeserializer) this.deserializer()).configured()) return false;
```

mjsax replied:
Would it be better to extract to logic into the test but not the Serde and (De)Serilalizer, but let the `Serde` just return `serializerConfigured()` and `deserializerConfigure()` (instead of `configured()`) ?

tedyu committed:
Address guozhang's review comments

tedyu commented:
@guozhangwang @mjsax 
Please let me know what else I need to do.

mjsax commented on clients/src/main/java/org/apache/kafka/common/serialization/Serdes.java:
```diff
@@ -38,6 +38,9 @@ public WrapperSerde(Serializer<T> serializer, Deserializer<T> deserializer) {
 
         @Override
         public void configure(Map<String, ?> configs, boolean isKey) {
+            if (configs == null) {
```
Not sure if we should add this -- a wrapper (de)serializer should do this check is appropriate -- also, a (de)serializer might need to do some initialization that does not depend on the config. Thus, we should forward the call and just pass through `null` IMHO.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java:
```diff
@@ -44,6 +44,14 @@
         this.partitioner = partitioner;
     }
 
+    public Serializer getKeySer() {
```
nit: `getKeySerde()` -- similar below

tedyu replied:
Since only Serializer is returned, I didn't include 'de' in the method name.

mjsax replied:
I see. Than I would prefer `getKeySerializer()` (it's a nit, but I like to avoid abbreviations)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -634,6 +636,26 @@ private KafkaStreams(final InternalTopologyBuilder internalTopologyBuilder,
     }
 
+    @SuppressWarnings("unchecked")
+    private void configureSerDe(final Set<SinkNode> sinks, final Set<SourceNode> sources) {
```
nit: `configureSerdes()` (to align naming with other classes/methods)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java:
```diff
@@ -115,6 +115,14 @@ public void init(final ProcessorContext context,
 
     @SuppressWarnings("unchecked")
     void initStoreSerde(final ProcessorContext context) {
+        if (context.appConfigs() != null) {
```
`appConfigs()` will never return `null` -- we can skip this test IMHO

mjsax replied:
It might also be nice, to do the check if `keySerde != null` only once and avoid a second check when creating `StateSerdes`. Something like:
```
void initStoreSerde(final ProcessorContext context) {
  final Serde<K> usedKeySerde;
    final Serde<V> usedValueSerde;
    if (keySerde == null) {
        usedKeySerde = (Serde<K>) context.keySerde();
     } else {
        usedKeySerde = keySerde;
        usedKeySerde.configure(context.appConfigs(), true);
    }
    // same for valueSerde
    serdes = new StateSerdes<>(
        ProcessorStateManager.storeChangelogTopic(context.applicationId(), name()),
        usedKeySerde,
        usedValueSerde);
```

guozhangwang replied:
About `appConfigs()` check: is it possible in some unit tests, that mock context is used which could return null? If not, then I agree it is not necessary to check `appConfigs() != null` anywhere.

mjsax replied:
Might be better to update the tests instead?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -476,7 +476,63 @@ public void shouldNotGetTaskWithKeyAndPartitionerWhenNotRunning() {
+            if (!((MyStringDeserializer) this.deserializer()).configured()) return false;
+            return true;
+        }
+    }
```
nit: add empty line below

mjsax commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -476,7 +476,63 @@ public void shouldNotGetTaskWithKeyAndPartitionerWhenNotRunning() {
+            return !called || configured;
+        }
+    }
+    static class MyStringDeserializer extends StringDeserializer {
```
nit: add empty line between both classes

mjsax commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -476,7 +476,63 @@ public void shouldNotGetTaskWithKeyAndPartitionerWhenNotRunning() {
+            super.configure(configs, isKey);
+            configured = true;
+        }
+        @Override
```
nit: empty line

mjsax commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -476,7 +476,63 @@ public void shouldNotGetTaskWithKeyAndPartitionerWhenNotRunning() {
+        public void configure(final Map<String, ?> configs, final boolean isKey) {
+            super.configure(configs, isKey);
+        }
+        public boolean configured() {
```
nit: empty line

mjsax reviewed (commented):
Thanks for the update @tedyu!

I would recommend to not extend any existing tests, but add a completely new test: `AutoSerdeConfigurationTest.java` -- the test should use different topologies and test `Consumed`, `Produced`, `Grouped`, `Materialized` and also `Processor API` topologies.

tedyu commented:
w.r.t. "Would it be better to extract to logic", I think the current form is more intuitive - serializer / deserializer may be called individually or both.

tedyu committed:
Address Matthias' comments

tedyu commented:
w.r.t. test composition, standalone test would copy topologies from existing tests.
I am not sure that is better than current form.

tedyu committed:
rename methods

tedyu committed:
add assertion for Consumed

tedyu committed:
Add assertion for Materialized

tedyu commented:
LeaderEpochIntegrationTest failure was not related to the changes.

tedyu commented:
retest this please

guozhangwang commented:
> w.r.t. test composition, standalone test would copy topologies from existing tests.

I'd encourage to have decomposed tests as well, although it introduced some duplicate code for setting up: in our past experience we have difficulties trouble shooting a failed test if it is a "compound" case covering many scenarios, and in the end the community decides to prefer modular tests.

guozhangwang commented:
LGTM modulo @mjsax 's comments (btw I agree with him that `configs == null` check is not necessary, as for `appConfig != null` if unit tests are not impacted then I also agree it can be removed).

@mjsax can merge after all comments are addressed.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -476,7 +476,67 @@ public void shouldNotGetTaskWithKeyAndPartitionerWhenNotRunning() {
+        }
+
+        public boolean configured() {
+            if (!((MyStringSerializer) this.serializer()).configured()) return false;
```
This can be simplified to 
```java
return ((MyStringSerializer) serializer()).configured() &&
                ((MyStringDeserializer) deserializer()).configured();
```

bbejeck reviewed (commented):
Thanks for the PR @tedyu.  Overall this looks good, and so far i only have one comment.

>I would recommend to not extend any existing tests, but add a completely new test: AutoSerdeConfigurationTest.java -- the test should use different topologies and test Consumed, Produced, Grouped, Materialized and also Processor API topologies.

I agree with @mjsax here, as it's better to have a separate test dedicated to testing one specific functionality vs. extending an existing test. 
This practice tends to make maintenance more manageable as the new functionality needs to change or gets updated then only the specific tests related to new features need to get updated vs. different tests scattered through the test folder.

tedyu commented:
I have added mock in related tests.

null check for config has been dropped.

tedyu committed:
Create KStreamSerDesIntegrationTest

tedyu committed:
Use mock in tests and remove null check for config

tedyu commented:
@mjsax @guozhangwang 
Do you have more comment ?

mjsax commented on docs/streams/developer-guide/dsl-api.html:
```diff
@@ -3548,6 +3548,7 @@ <h4><a id="streams_concepts_globalktable" href="#streams_concepts_globalktable">
 clicksPerRegion.toStream.to(outputTopic)
               </pre>
               <p>A complete example of user-defined SerDes can be found in a test class within the library.</p>
+              <p>The auto-configure of user-defined SerDes would be done at KafkaStreams construction time.</p>
```
Can we add something similar to `datatypes.html` -- there is a sections about Serdes.

tedyu replied:
Are you talking about the section "configuring-serdes" ?

Please let me know whether the addition should be made in datatypes.html or here.

mjsax replied:
I think it's best to add both updates. Here and in `datatypes.html`. Maybe we can add a sentence to both sections "CONFIGURING SERDES" and "OVERRIDING DEFAULT SERDES"

mjsax replied:
As above. Not create new paragraph but add sentence to existing one:
```
<p>A complete example of user-defined SerDes can be found in a test class within the library.
   Kafka Streams will auto-configure the passed-in SerDes objects, i.e., you don't need to call <code>configure()</code> manually.</p>
```

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java:
```diff
@@ -115,10 +115,25 @@ public void init(final ProcessorContext context,
             ProcessorStateManager.storeChangelogTopic(context.applicationId(), name()),
-            keySerde == null ? (Serde<K>) context.keySerde() : keySerde,
-            valueSerde == null ? (Serde<V>) context.valueSerde() : valueSerde);
+                usedKeySerde,
```
nit: fix indention

tedyu replied:
usedKeySerde is indented 4 spaces. Isn't that correct?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredSessionStore.java:
```diff
@@ -69,6 +69,13 @@
+        }
+        if (valueSerde != null) {
+            valueSerde.configure(conf, false);
+        }
```
I think this should be the same code as in `MeteredKeyValueStore` -- also `MeteredTimestampedKeyValueStore` below and `MeteredTimestampedWindowStore`, and `MeteredWindowStore` -- maybe it's worth to extract a shared method that all 5 `MeteredXxxStores` are using? We could add a `MeteredStoreUtils` with a static method?

Just updating each store individually might also be fine.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KStreamSerDesIntegrationTest.java:
```diff
@@ -0,0 +1,491 @@
+ * by virtue of having a large commit interval
+ */
+@Category({IntegrationTest.class})
+public class KStreamSerDesIntegrationTest {
```
Thanks for adding a new test.

I think we can further improve this test. Atm, it seem to be copy of another test, instead of tailored test to the Serde-auto-configuring feature. Including proper test method names.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KStreamSerDesIntegrationTest.java:
```diff
@@ -0,0 +1,491 @@
+    }
+    @Test
+    @SuppressWarnings("unchecked")
+    public void shouldSendCorrectResults_NO_OPTIMIZATION() throws Exception {
```
In this test, we don't care about correct result if optimization is disable -- in general we want to test Serde-auto-config feature. Test method should be renamed accordingly.

There are multiple ways to structure the test and break it into smaller pieces. Not sure atm what the best way is, but it should be easy to read and understand. Maybe we can have one test for Processor API, with one source, one stateful processor and one sink?

And one more test for the DSL? We should also cover the global state/KTable case.

tedyu replied:
Since it would take some time to add test coverage / properly structure the tests, can the refinement of the integration test be done in a follow-up PR ?

tedyu replied:
I simplified KStreamSerDesIntegrationTest by removing irrelevant test and merging test cases.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/KStreamSerDesIntegrationTest.java:
```diff
@@ -0,0 +1,491 @@
+            return !called || configured;
+        }
+    }
+    static class MyIntDeserializer extends IntegerDeserializer {
```
Can we update the test to only use one data type? Don't think we gain any test coverage using multiple -- this seems to be distraction what the purpose of the test is.

mjsax reviewed (commented):
Thanks for the update @tedyu!

Some more comments on improving the added test.

tedyu committed:
Address Matthias' review comments on SerDe

tedyu commented:
I addressed the comments about usedKeySerde in latest commit.
I didn't extract the code into helper method since special casting is required for two of the SerDe's.

tedyu committed:
Add description in datatypes.html

tedyu committed:
Simplify KStreamSerDesIntegrationTest

tedyu committed:
Modify TableTableJoinIntegrationTest with custom SerDe and assertion for configuration

tedyu committed:
Fix compilation warning

tedyu commented:
w.r.t. global state, I found the following two tests:

streams/src/test/java/org/apache/kafka/streams/integration/KTableSourceTopicRestartIntegrationTest.java
streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java

shouldRestoreStateFromSourceTopic and shouldRestoreStateFromChangelogTopic from RestoreIntegrationTest seem to qualify.
Please let me know if plugging in custom SerDe in the above two sub-tests is the way to go.

tedyu committed:
Add SerDe assertion in RestoreIntegrationTest

tedyu committed:
Address checkstyle warning

tedyu commented:
@mjsax 
Can you take another look ?

thanks

tedyu commented:
ping @mjsax

mjsax commented on docs/streams/developer-guide/datatypes.html:
```diff
@@ -69,6 +69,7 @@ <h2>Configuring SerDes<a class="headerlink" href="#configuring-serdes" title="Pe
     </div>
     <div class="section" id="overriding-default-serdes">
       <h2>Overriding default SerDes<a class="headerlink" href="#overriding-default-serdes" title="Permalink to this headline"></a></h2>
+      <p>The auto-configure of user-defined SerDes is done at KafkaStreams construction time.</p>
```
Nit: No need to start a new paragraph. `<p>` -- we can put both sentences into one (note, I would like to fix `serde settings` -> `SerDe settings` in the first/existing sentence):
```
<p>You can also specify SerDes explicitly by passing them to the appropriate API methods, which overrides the default SerDe settings.
   For this case, Kafka Streams will auto-configure the passed-in SerDes objects, i.e., you don't need to call <code>configure()</code> manually.</p>

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/TableTableJoinIntegrationTest.java:
```diff
@@ -63,9 +67,71 @@ public void prepareTopology() throws InterruptedException {
     final private String expectedFinalMultiJoinResult = "D-d-d";
     final private String storeName = appID + "-store";
 
+    static class MyLongSerializer extends LongSerializer {
```
We should not test auto-configure feature within this test at all.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java:
```diff
@@ -116,6 +116,7 @@ public void shutdown() {
     }
 
     @Test
+    @SuppressWarnings("unchecked")
```
Auto-configuration has nothing to do with state restoring and testing auto-configuration should be be included in this test at all.

mjsax reviewed (commented):
Sorry for the long delay. Some nit comments about the doc changes.

I was thinking about testing once more. It seems that we don't really need an integration test, but unit testing should be sufficient. Hence, adding new test methods to `MeteredKeyValueStoreTest` etc, ie, for all 5 stores as well as `KafkaStreamsTest` should be sufficient. For `KafkaStreamsTest`, we can use a topology like:
```
builder
    .stream(... Consumed.with())
    .through(...)
    .selectKey()
    .groupByKey()
    .count()
    .toStream().to(...,Produced.With())`;
builder.globalTable(..., Consumed.with());
```
and only call `new KafkaStreams().close()` -- there is no need to actually run the program to test if auto-configuring works. Hence, we don't need to use `EmbeddedKafkaCluster` etc and can simplify the test.

Does this make sense? If you are not sure, I could also write the tests myself in a follow up PR. For this case, just revert unnecessary changes to test code so we could merge this PR.

tedyu committed:
Address Matthias' comments

tedyu commented:
Thanks for the review, @mjsax 

I have modified the doc changes and removed the new integration test.

Since you have better idea on how the test should be formulated, it would be nice if you or someone else can implement the test in a follow-on PR.

tedyu commented:
testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup failure was not related.

tedyu commented:
retest this please

tedyu commented:
retest this please

tedyu commented:
@mjsax 
Do you have other comment ?

mjsax commented:
Thanks @tedyu! This LGTM. I would like to first write the unit test based on this PR, and if they pass, we can merge both PR one after each other. Just as save guard to not merge anything broken (even if I think this PR is fine). May take a couple of day until I find time. I keep you posted.

mjsax commented:
@tedyu Follow up with unit-test: https://github.com/apache/kafka/pull/6610 -- please take a look.

tedyu committed:
Address Guozhang's comment

mjsax commented:
Thanks for the PR @tedyu. Merged to `trunk`.
